binary-option/program/src/processor.rs

```diff
@@ -0,0 +1,725 @@
+                &escrow_account_info,
+                &seller_account_info,
+                &authority_info,
+                n_s * buy_price,
+                seeds,
+            )?;
+            spl_token_transfer_signed(
+                &token_program_info,
+                &escrow_account_info,
+                &buyer_account_info,
+                &authority_info,
+                n * sell_price,
+                seeds,
+            )?;
+            binary_option.decrement_supply(n_s)?;
+        }
+        /*
+        When n is greater than n_s bust less than n_b, this means that the seller has put on a position that is different from their
+        existing position, and the buyer has reduced their inventory. We will burn and mint tokens such the buyer's net change in
+        position is +n and the seller's net change is -n. Both parties are also entitled to the locked up funds for their positions that were closed.
+        The net change in tokens can be calculated as follows: (-n - n_b + n - n_b) / 2 = -n_b. This always results in a decrease in total
+        circulation.
+        */
+        [false, true] => {
+            msg!("Case 4");
+            spl_burn(
+                &token_program_info,
+                &seller_long_token_account_info,
+                &long_token_mint_info,
+                &seller_info,
+                n,
+            )?;
+            spl_burn(
+                &token_program_info,
+                &buyer_short_token_account_info,
+                &short_token_mint_info,
+                &buyer_info,
+                n_b,
+            )?;
+            b_s -= n_b;
+            s_l -= n;
+            spl_mint_to(
+                &token_program_info,
+                &buyer_long_token_account_info,
+                &long_token_mint_info,
+                &authority_info,
+                n - n_b,
+                seeds,
+            )?;
+            b_l += n - n_b;
+            spl_token_transfer(
+                &token_program_info,
+                &buyer_account_info,
+                &escrow_account_info,
+                &buyer_info,
+                (n - n_b) * buy_price,
+            )?;
+            spl_token_transfer_signed(
+                &token_program_info,
+                &escrow_account_info,
+                &buyer_account_info,
+                &authority_info,
+                n_b * sell_price,
+                seeds,
+            )?;
+            spl_token_transfer_signed(
+                &token_program_info,
+                &escrow_account_info,
+                &seller_account_info,
+                &authority_info,
+                n * buy_price,
+                seeds,
+            )?;
+            binary_option.decrement_supply(n_b)?;
+        }
+    }
+    // Delegate the burn authority to the PDA, so a private key is unnecessary on collection
+    // This can probably be optimized to reduce the number of instructions needed at some point
```

Comment:
How else can Collect be performed by any user? This is only to give the program the ability to burn the user's tokens to prevent them from double collecting.

Reply:
Oh I see what you mean, hm, it'll be tough to make the `Collect` instruction permissionless.  To be honest, for the first round I'd have it *not* be permissionless, so people have to collect the money themselves.  Otherwise, you could store a separate record for each person and not use tokens, but that breaks the whole "use this in DeFi" model.